### ISSUE TITLE
Add `asAsyncData` helper method for fluent Promises (e.g., Postgrest-js) integration

### ISSUE THREAD

**Issue by leynier** [nuxt-modules/supabase#443]
### Is your feature request related to a problem? Please describe.  
When using libraries like `postgrest-js` in a Nuxt project, the returned objects (e.g., `PostgrestBuilder` or `PostgrestFilterBuilder`) implement `PromiseLike`. While this allows resolving results with `await` or `.then()`, the preferred Nuxt pattern for handling asynchronous data is through the `useAsyncData` composable.  

Currently, manually wrapping these fluent chains in `useAsyncData` can feel verbose and repetitive, reducing code clarity and developer experience.

---

### Describe the solution you'd like  
It would be helpful to introduce a helper method like `asAsyncData()` directly on `PromiseLike` objects (e.g., `PostgrestBuilder` in postgrest-js). This method would wrap the current fluent chain into `useAsyncData`, enabling a more concise and fluent syntax.

**Example:**

Instead of:

```ts
const { data, status, error } = await useAsyncData(
  'fetch-user',
  () => supabase.from('users').select('*').eq('id', 1)
);
```

You could write:

```ts
const { data, status, error } = await supabase
  .from('users')
  .select('*')
  .eq('id', 1)
  .asAsyncData('fetch-user');
```

---

### Describe alternatives you've considered  
- **Manual wrapping**: Continue manually wrapping each fluent chain in `useAsyncData`. While functional, this approach leads to repetitive code and reduces readability.  
- **Utility functions**: Creating a standalone helper function to wrap the promises. However, this doesn't integrate cleanly into the fluent interface pattern of libraries like `postgrest-js`.

---

### Additional context  
The solution can leverage **prototype chaining** to extend the behavior of `PromiseLike` objects such as `PostgrestBuilder`, and **declaration merging** to ensure proper TypeScript type support. This approach avoids modifying the original library code and provides a clean, composable integration into Nuxt.  

Related Nuxt documentation: [useAsyncData](https://nuxt.com/docs/getting-started/data-fetching).

**Comment by Nuzzlet** [comment]
I'm interested in taking this on. I've written something almost identical for one of my projects.  I think there needs to be some discussion about the correct approach. I've made a [fork](https://github.com/Nuzzlet/supabase) to hold my changes as I figure this out.

To me, the goals of this would be as follows:

- Simpler API for cleaner code
- Re-use postgrest-js as much as possible
- Work with custom schemas
- Allow passing filters as a computed function, so re-fetching can occur if the query changes
- Auto cache / key based on the fetch URL so identical queries can share results
- Support useAsyncData options such as server, immediate, and watch
- Fully typed, including errors (since errors should always be Postgrest Errors)
- _Offer an abstraction around range() for pagination style, and infinite scroll style loading_

To me, the way to implement this is with a define from and select statement, and with filters => {} as computed function.

In your example:

> `const { data, status, error } = await supabase
  .from('users')
  .select('*')
  .eq('id', 1)
  .asAsyncData('fetch-user');`

I think this is a bad idea, as it would be modifying the supabase postgres-js code. To me it makes more sense as a new function, something like this:

`const { data } = await useSupabaseQuery('artists', `*, ratings, photos!inner(*), reviews(*)`, (filter) => {
  filter.limit(3, { referencedTable: 'reviews' }).eq('id', route.params.artist)
  return filter.single()
})`

One of the big issues is the filter _(PostgrestFilterBuilder)_ has methods such as throwsOnError() which would likely cause issues. .single() is okay because it modifies the exported type which we can capture with a generic... which my fork does.

**Comment by dan-hale** [comment]
I ended up creating my own module to do this. Check it out [here](https://github.com/dan-hale/nuxt-supabase-query)

**Comment by larbish** [comment]
Hello @leynier, @dan-hale 

Thank you both for your contributions and for taking the time to propose this update.

The update you're proposing seems a bit overkill to me and by removing the `useAsyncData` we're loosing all advantages that comes with it: https://nuxt.com/docs/api/composables/use-async-data#params (including `useNuxtData` and `refreshNuxtData`. 

Considering the added complexity this update introduces, I do not feel confident integrating and maintaining it within the module. If this approach provides value for your specific project, I recommend implementing it directly in your application code. My goal is to keep the module as simple and intuitive as possible, ensuring it remains an easy way to integrate Supabase with Nuxt while only adhering (without adding any extra complexity) to the best practices each technology has to offer.

Thank you again for your suggestion and understanding! 🙏

**Comment by dan-hale** [comment]
> Hello @leynier, @dan-hale 
> 
> Thank you both for your contributions and for taking the time to propose this update.
> 
> The update you're proposing seems a bit overkill to me and by removing the `useAsyncData` we're loosing all advantages that comes with it: https://nuxt.com/docs/api/composables/use-async-data#params (including `useNuxtData` and `refreshNuxtData`. 
> 
> Considering the added complexity this update introduces, I do not feel confident integrating and maintaining it within the module. If this approach provides value for your specific project, I recommend implementing it directly in your application code. My goal is to keep the module as simple and intuitive as possible, ensuring it remains an easy way to integrate Supabase with Nuxt while only adhering (without adding any extra complexity) to the best practices each technology has to offer.
> 
> Thank you again for your suggestion and understanding! 🙏

I agree with this conclusion. I tried a few different approaches and all of theme seemed like compromises. The slight added connivence is not worth the tradeoffs.

**Comment by leynier** [comment]
Hi @larbish 

Following up on the previous discussion about integrating a method like `asAsyncData()` to simplify using `PostgrestBuilder` with `useAsyncData`, I decided to implement this functionality as a separate module for my own project (https://github.com/leynier/nuxt-supabase-query-as-async-data). The goal was to improve the developer experience when working with Supabase queries in Nuxt, while preserving the method chaining fluidity of `postgrest-js` and leveraging the benefits of `useAsyncData`.

### What does this module do?

This module extends the prototypes of both `PostgrestBuilder` instances and functions returning them by adding a method named `asAsyncData()`. This method allows wrapping Supabase queries directly in `useAsyncData` in a fluent and concise way.

#### Example usage:

Instead of writing:

```ts
const { data, error } = await useAsyncData(
  'fetch-user',
  () => supabase.from('users').select('*').eq('id', 1)
);
```

With this module, you can write:

```ts
const { data, error } = await supabase
  .from('users')
  .select('*')
  .eq('id', 1)
  .asAsyncData('fetch-user');
```

### How does it work?

1. **Prototype extension:**  
   It uses prototype chaining to add the `asAsyncData()` method to both `PostgrestBuilder` objects and functions returning them.

2. **TypeScript support:**  
   It uses declaration merging to fully type the `asAsyncData()` method, including support for `PostgrestSingleResponse` and error handling.

3. **Compatibility with `useAsyncData`:**  
   The method respects `useAsyncData` options like `server`, `immediate`, and `watch`, maintaining Nuxt best practices.

### Module code

The module is composed of two main files:

1. **plugin.ts** — adds the `asAsyncData()` method via prototype extension.
2. **types.d.ts** — adds typings to ensure full TypeScript compatibility.

#### plugin.ts

```ts
import { PostgrestClient } from "@supabase/postgrest-js";
import { defineNuxtPlugin, useAsyncData } from "nuxt/app";

export default defineNuxtPlugin((_) => {
    const supabase = new PostgrestClient("https://dummy.com");
    const tempQueryFunction = () => supabase.from("_temp").select("*");
    const queryProtoFunction = Object.getPrototypeOf(tempQueryFunction);
    if (queryProtoFunction && !queryProtoFunction.asAsyncData) {
        queryProtoFunction.asAsyncData = function (key: string, options = {}) {
            return useAsyncData(key, () => this(), options);
        };
    }
    const tempQueryObject = supabase.from("_temp").select("*");
    const queryProtoObject = Object.getPrototypeOf(tempQueryObject);
    if (queryProtoObject && !queryProtoObject.asAsyncData) {
        queryProtoObject.asAsyncData = function (key: string, options = {}) {
            return useAsyncData(key, () => this, options);
        };
    }
});
```

#### types.d.ts

```ts
import "@supabase/postgrest-js";
import type { PostgrestBuilder, PostgrestSingleResponse } from "@supabase/postgrest-js";
import type { AsyncData, AsyncDataOptions } from "nuxt/app";

declare module "@supabase/postgrest-js" {
  interface PostgrestBuilder<Result> {
    asAsyncData(key: string, options?: AsyncDataOptions<PostgrestSingleResponse<Result>>): AsyncData<PostgrestSingleResponse<Result>, Error>;
  }
}

declare global {
    interface Function {
        asAsyncData<Result>(this: () => PostgrestBuilder<Result>, key: string, options?: AsyncDataOptions<PostgrestSingleResponse<Result>>): AsyncData<PostgrestSingleResponse<Result>, Error>;
    }
}
```

### Proposal

If this functionality sounds useful, I’d love to see it integrated directly into the `@nuxtjs/supabase` module. It would offer developers a smoother and more concise API when working with Supabase queries in Nuxt, without requiring them to re-implement the same logic repeatedly.

I’m open to feedback or suggestions to align this approach with the standards and vision of the module.

Thanks for your time and consideration! 🙌
